### ISSUE TITLE
[Security] Some thoughts

### ISSUE THREAD

**Issue by ro0NL** [symfony/symfony#42667]
Hi,

(follow-up of #42650)

I believe our main goal for the new "security system" was to have always authenticated tokens, isnt it? As such i feel like `make TokenInterface::getUser() nullable` swaps a technical issue for a (severe) semantical one, defeating our initial goal. I want to verify it's 100% intentional.

https://github.com/symfony/symfony/blob/bf8ecc417fb900db91978cf6ec933319f7070083/src/Symfony/Component/Security/Http/Authenticator/AuthenticatorInterface.php#L73

I'd be able to bypass the passport user, and provide an unauthenticated token isnt it? Meaning i'm effectively re-introducing a AnonymousToken. But anonymous means token=null :thinking:

https://github.com/symfony/symfony/blob/e3f72bdb55b23e5b09cf66cf203c32c8d9e90ac8/src/Symfony/Component/Security/Core/Authentication/Token/AnonymousToken.php#L21-L23

So, IIUC, we're now back to `$loggedIn = $token && $token->getUser()`, rather than our initial intention `$loggedIn = !!$token` :) It looks like we're back at sqaure 1.

Confirmed here:

https://github.com/symfony/symfony/blob/bf8ecc417fb900db91978cf6ec933319f7070083/src/Symfony/Component/Security/Core/Authentication/AuthenticationTrustResolver.php#L25-L27

I dont think the check itself should require a service layer. Meaning we probably need better semantics for `AuthenticationTrustResolver::isAuthenticated` as well, or remove it. (btw calling it without args to me implies "the current token")

Now, to get back to "tokens are always authenticated" AND fix our technical issue in the voter subsystem i'd like to propose these alternative maybe-possible solutions;

1) explore  eg. `AnonymousUser / NullUser`
2) introduce some signaling exception from getUser
3) leverage some role name
4) `vote(?TokenInterface $token)` one way, or another :}

More side-related questions:

- how do we explain `TokenInterface::getRoleNames` vs `UserInterface::getRoles`, in both the current and previous architecture? What if i have a token with roles but no user currently?
- is `CredentialsInterface` a pre-mature abstraction? Do we have another sementical issue given i could do the same using `SelfValidatingPassport` + badges?
- Do we really need `AuthenticationTrustResolverInterface`? rather than some static api.

Thanks :)

cc @chalasr, @nicolas-grekas , @wouterj also

**Comment by nicolas-grekas** [comment]
There's some expectation mismatch that this is ignoring and led to the nullable state: ppl want to be able to vote on non-authenticated tokens, to control what public access means for their app.

I agree that `AuthenticationTrustResolver::isAuthenticated` might not be needed anymore.

**Comment by ro0NL** [comment]
Im not ignoring it :) i'm wondering if the current solution is really really worth it, since our goal was "to have always authenticated tokens".

**Comment by wouterj** [comment]
> I believe our main goal for the new "security system" was to have always authenticated tokens, isnt it?

Yes

> I'd be able to bypass the passport user, and provide an unauthenticated token isnt it?

Nullable user is only needed in the `AccessDecisionManager` (and voters) due to the reason @nicolas-grekas just mentioned. I'm :+1: if you submit a PR to throw a logic exception if user is `null` in the token returned by `createAuthenticatedToken()`/`createToken()`.

> But anonymous means token=null

I guess you meant `user=null` here, but to be absolutely clear: `token = null` *is* the new "anonymous".

> Meaning we probably need better semantics for `AuthenticationTrustResolver::isAuthenticated` as well, or remove it.

I agree, I would accept a PR that removes this method.

> (btw calling it without args to me implies "the current token")

It would have been nice if we could do it like this, but we have to be consistent with the other trust resolver method, which don't behave like this if you don't pass an argument (they probably used `= null` to mark the argument nullable back in the PHP 5 days).

> `vote(?TokenInterface $token)` one way, or another :}

Yes, but no. Backwards compatibility sometimes forces you to make a minimal concession on design, in favor of stability and BC (btw, I 100% disagree that adding a conditional `null` here is a "severe semantical issue").

---

Please ask your unrelated questions in a different issue. It's hard to follow github issues with multiple discussions

**Comment by ro0NL** [comment]
> Nullable user is only needed in the AccessDecisionManager (and voters) 

but it's leaking everywhere now.

> token = null is the new "anonymous".

and user=null means?

**Comment by wouterj** [comment]
Actually, those unrelated questions have already been answered:

> is `CredentialsInterface` a pre-mature abstraction? Do we have another sementical issue given i could do the same using `SelfValidatingPassport` + badges?

https://github.com/symfony/symfony/issues/39302

> Do we really need AuthenticationTrustResolverInterface? rather than some static api.

https://github.com/symfony/symfony/pull/35860

**Comment by ro0NL** [comment]
i don't think it answers my questions. Sry, should have created a discussion at this point probably, but it's fairly new :)

**Comment by wouterj** [comment]
Other idea: introduce a new `AccessDecisionToken` that decorates either the real authenticated token or no token (public access). This means we can have a special token API for voters and don't impact the application logic.

**Comment by ro0NL** [comment]
Now we're talking. Decoration sounds promising.

**Comment by chalasr** [comment]
Sounds good to me as well

**Comment by ro0NL** [comment]
Another way to look at this is `@return UserInterface <a bunch of explaining text here>` while keeping nullablity. But im not really fond of inventing our own laws.

**Comment by ro0NL** [comment]
Now closing btw in favor of general topic :) No intention to push _something_.

**Comment by wouterj** [comment]
In case anyone wants to actually try the decorator change, some dummy sketch code

```php
interface TokenInterface
{
    /* @return UserInterface (string|\Stringify in <6.0 for BC) */
    public function getUser();

    // ...
}

// Maybe remove TokenInterface implements in 6.0 and change voter to vote(AccessDecisionToken $token) ?
class AccessDecisionToken implements TokenInterface
{
    private $token;

    public function __construct(?TokenInterface $token)
    {
        $this->token = $token;
    }

    public function isAuthenticated(): bool
    {
        return null !== $this->token;
    }

    public function getToken(): TokenInterface
    {
        if (null === $this->token) {
            throw new \LogicException(sprintf('Cannot call "%1$s::getToken()" if "%1$s::isAuthenticated()" returned false.', __CLASS__));
        }

        return $this->token;
    }

    public function getUser():
    {
        trigger_deprecation('symfony/security-core', '5.4', 'Calling "%s()" is deprecated.', __METHOD__);

        return '';
    }
    
    // ...
}
```

**Comment by ro0NL** [comment]
@wouterj maybe transform to discussion? can you?

The fundamental question is i believe, to have always authenticated tokens, how do we envision anon. token voting. This is the part where i don't have time to digest further.

If anything, i found `$loggedIn = !!$token` a super noble goal :)

**Comment by ro0NL** [comment]
Another POV, maybe as of 7.x we are only talking passports to users :D time will tell.

**Comment by wouterj** [comment]
> The fundamental question is i believe, to have always authenticated tokens, how do we envision anon. token voting. This is the part where i don't have time to digest further.

The main thing is: how to have anon. token voting with the current method signature (which we can't change).

Imho, either the current way or by having a specialized token API for voters. We already create a special token for anon. just within the scope of voters in [`AuthorizationChecker::isGranted()`](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php#L77) we can quite easily change this by a decorated token (e.g. using my snippet above).

**Comment by ro0NL** [comment]
>  with the current method signature

or fresh API design at the namespace level :}
